### ISSUE TITLE
Is there any point anymore in keeping WebGL1 support?

### ISSUE THREAD

**Issue by wizgrav** [mrdoob/three.js#22689]
I thought it was a nice issue to raise. WebGL2 as minimum would make a lot of things better in the library, for instance with the bone textures the glsl code becomes way nicer and requires one less uniform, because we can read texture sizes directly from the shader. I suspect this may apply elsewhere as well. Is there still userbase out there that doesn't support WebGL2? It's like almost a decade old mobile technology

**Comment by mrdoob** [comment]
I would love to get rid of all the webgl1 code but I feel it would be premature.

What I was thinking was that we could remove the automatic WebGL1 fallback and add a `enableWebGL1Fallback` parameter to `WebGLRenderer`'s constructor. That way we'll hear all the cases while still being able to have a solution.

And at that point we'd be able to have a better feeling of when to remove all the WebGL1 code path.

**Comment by donmccurdy** [comment]
Note that having the latest Safari or Chrome does not necessarily backport WebGL 2 support to older devices – you need both an updated browser AND a compatible device. I think it's very likely (but not sure how to test this...) that low-cost Android devices being manufactured _today_ and running the latest version of Chrome may still not support WebGL 2.0. Unfortunately the statistics on https://caniuse.com/?search=webgl2 are meaningless in this case, they just indicate that the browser version is common, not that device support is common.

**Comment by wizgrav** [comment]
@donmccurdy yeah that's valid, though I suspect that there are not many of these devices in circulation. Having it disabled in a sizable chunk of ios devices is the big problem though. Asking the user to enable it takes 5 or more taps so not viable. The browser vendors could gather the stats so we know where we stand

**Comment by vanruesc** [comment]
> Definitely stats would be really helpful here though.

I think the next best official stats would be the [Android dashboard](https://developer.android.com/about/dashboards#OpenGL) which states that 8.5% of all devices only support OpenGL ES 2.0 as of July 31, 2021. This at least implies no support for Opengl 3.0+ / WebGL 2.

**Comment by mrdoob** [comment]
@vanruesc Interesting...

For iOS land we can probably use [this page](https://developer.apple.com/support/app-store/). 
The numbers are from June 2021 though.

**Comment by jeremy-coleman** [comment]
I am probably in the top 0.1 % of household income globally and my phone (iphone 6) doesnt support webgl2. Dont be ridiculous.

**Comment by donmccurdy** [comment]
@jeremy-coleman out of curiosity – my iPhone SE (2016) supposedly has the same internal components as an iPhone 6s, but does indeed support WebGL 2 (after this year's iOS / Safari updates) without a flag. Is that not the case on the latest iOS for you? I guess that would tell us the exact year and model where WebGL 2 support ends, if not.

**Comment by jeremy-coleman** [comment]
I am not the target cutoff. Schools in third world countries using three for sci vis are.

**Comment by donmccurdy** [comment]
Sure, the intention is to base timing on device statistics (globally, if possible) and not any particular individual's phone. Please do presume good intent here — we're discussing global statistics in the thread above and @mrdoob has already commented "it would be premature" at this stage. I don't think it's the right time to drop WebGL 1.0 support yet either. The goal at this point is how to find statistics that would allow us to make an informed decision on (eventual) timing.

**Comment by jeremy-coleman** [comment]
Hmm, got this yesterday and remembered this thread. ota driver update i guess

![2ED75BA3-103E-4564-ACFD-A78B34BB418F](https://user-images.githubusercontent.com/25135855/144748023-41ad59fa-71f4-4a1b-bb09-8f9ca6fb7778.png)

**Comment by jeremy-coleman** [comment]
> @jeremy-coleman out of curiosity – my iPhone SE (2016) supposedly has the same internal specs as an iPhone 6s, but does indeed support WebGL 2 (after this year's iOS / Safari updates) without a flag. Is that not the case on the latest iOS for you? I guess that would tell us the exact year and model where WebGL 2 support ends, if not.

Follow up, im dumb, my wife told me i have an iphone 8

**Comment by tangobravo** [comment]
WebGL2 is enabled by default from iOS 15, but has a plethora of different bugs on the different versions of iOS. 15.3 was looking pretty good, but 15.4 seems to have a number of big regressions.

Right now for our content we're considering enforcing WebGL1 on iOS still until Apple at least manage a couple of releases without introducing serious regressions into WebGL2.
